test/integration/compose/application-manager.spec.ts

```diff
@@ -327,6 +329,370 @@ describe('compose/application-manager', () => {
+			};
+			const { currentServices, targetServices } =
+				await createCurrentAndTargetServicesWithLabels(labels);
+
+			/**
+			 * Before target image starts downloading, infer kill step
+			 */
+			const targetApps = createApps(
+				{
+					services: targetServices,
+				},
+				true,
+			);
+			const c1 = createCurrentState({
+				services: currentServices,
+				networks: [DEFAULT_NETWORK],
+			});
+			const steps1 = await applicationManager.inferNextSteps(
+				c1.currentApps,
+				targetApps,
+				{
+					downloading: c1.downloading,
+					// Pretend that images from current release have already
+					// been deleted from the image table
+					availableImages: [],
+					containerIdsByAppId: c1.containerIdsByAppId,
+				},
+			);
+			// Service `old` should be killed
+			const [killStep] = expectSteps('kill', steps1, 1);
+			expect(killStep)
+				.to.have.property('current')
+				.that.deep.includes({ serviceName: 'old' });
+			// No fetch step should be generated until kill completes
+			expectNoStep('fetch', steps1);
+
+			/**
+			 * While current services are stopping, infer noop step
+			 */
+			const c2 = createCurrentState({
+				services: currentServices.map((svc) => {
+					if (svc.serviceName === 'old') {
+						svc.status = 'Stopping';
+					}
+					return svc;
+				}),
+				networks: [DEFAULT_NETWORK],
+			});
+			const steps2 = await applicationManager.inferNextSteps(
+				c2.currentApps,
+				targetApps,
+				{
+					downloading: c2.downloading,
+					// Pretend that images from current release have already been
+					// deleted from the image table to avoid saveImage / removeImage steps
+					availableImages: [],
+					containerIdsByAppId: c2.containerIdsByAppId,
+				},
+			);
+			// Noop while service `old` is stopping
+			expectSteps('noop', steps2, 1);
+			// No fetch step should be generated until kill completes
+			expectNoStep('fetch', steps2);
+
+			/**
+			 * After current services not in target have stopped, infer fetch step
+			 */
+			const c3 = createCurrentState({
+				// Set current state to after `updateMetadata` step for `main` service
+				services: currentServices
+					.filter((svc) => svc.serviceName !== 'old')
+					.map((svc) => {
+						svc.releaseId = 2;
+						svc.commit = 'new-release';
+						return svc;
+					}),
+				networks: [DEFAULT_NETWORK],
+			});
+			const steps3 = await applicationManager.inferNextSteps(
+				c3.currentApps,
+				targetApps,
+				{
+					downloading: c3.downloading,
+					// Pretend that images from current release have already been
+					// deleted from the image table to avoid saveImage / removeImage steps
+					availableImages: [],
+					containerIdsByAppId: c3.containerIdsByAppId,
+				},
+			);
+			const [fetchStep] = expectSteps('fetch', steps3, 1);
+			expect(fetchStep).to.have.property('serviceName').that.equals('new');
+		});
+
+		it('should not infer a fetch step for target service B before current service A image removal with delete-then-download', async () => {
```

Comment:
Again, if A and B are different services, I'd expect the opposite. We don't want to remove the old service image until we are ready with the new release images.